yjit/src/codegen.rs

```diff
@@ -1326,6 +1326,48 @@ fn guard_object_is_array(
     asm.jne(side_exit.as_side_exit());
 }
 
+/// This guards that a special flag is not set on a hash.
+/// By passing a hash with this flag set as the last argument
+/// in a splat call, you can change the way keywords are handled
+/// to behave like ruby 2. We don't currently support this.
+fn guard_object_is_not_ruby2_keyword_hash(
+    asm: &mut Assembler,
+    object_opnd: Opnd,
+    side_exit: CodePtr,
+) {
+    asm.comment("guard object is not ruby2 keyword hash");
+
+    let label = asm.new_label("exit");
+    asm.test(object_opnd, (RUBY_IMMEDIATE_MASK as u64).into());
+    asm.jnz(label);
+
+    asm.cmp(object_opnd, Qfalse.into());
+    asm.je(label);
```

Comment:
I think you can reuse `guard_object_is_heap`. Also it's probably more readable to curve out the next check as `guard_object_is_hash` as well for consistency with the `string`/`array` counterparts.

Reply:
guard_object_is_heap isn't what I'm trying to do. If it is not a heap object, that's fine. I'm not going to exit.

The logic here is:
```python
if not(heap) object: continue
if heap && hash && ruby2keywords : exit
```

Reply:
Ah okay, that makes sense.